docs/saved_objects/saved_object_access_control_design_proposal.md

```diff
@@ -0,0 +1,216 @@
+
+This document discusses the proposed data access control for OpenSearch Dashboards, which is independent from the *multi-tenancy* feature provided by the security plugin. 
+
+Glossary
+
+* **Saved object**: Any data that is managed by OpenSearch Dashboards are called saved objects, such as index patterns, visualization and dashboard definition. In a common setup, each saved object is a document stored in `.opensearch_dashboards` index. In some context, it is also called library items, in #4615
+* **Workspace**: a logical container of OpenSearch Dashboards features and saved objects
+
+## Assumptions & Scopes
+
+* Authentication is out of the scope. We can assume the users are authenticated and user profile (user name and group info) has been injected into each request
+* Feature (API) access control is out of scope
+* Data (saved objects) access control is in scope
+
+## Design
+
+### Personas
+
+There are following personas in *Workspace* use cases:
+* **Super admin**: the administrator of the OpenSearch Dashboards, who has complete access to all functions and data of OpenSearch Dashboards.
+* **Workspace admin**: Workspace admin have complete access to workspace configurations and saved objects in the workspace. Workspace creator will become the workspace admin when a workspace is created. Workspace admin most time can be used exchangeably with and workspace owner. Workspace admin
+* **Workspace operator**: workspace operator has permission to view, create and update saved objects in the workspace.
+* **Workspace viewer**: workspace operator only has permission to view the saved objects in the workspace
+
+Please note that, the workspace admin, operator and viewer persona are specific to the workspace. e.g. one workspace operator in one workspace can be an admin in another workspace.
+
+super-admin is a stack level role, it can be configured in the `opensearch_dashboards.yml` config file.
+
+### Discretionary Access Control for Data Access Control
+
+In order to support allowing a resource owner to manage the access to the resource feature, we can employ discretionary access control model. The idea is that, for each OpenSearch Dashboards saved object, we can attach an ACL (access control list) to it to save the principals and permission to operate that saved object. When a user is attempting to perform any operation on that saved object, we will first of all evaluate the ACL and then allow or deny the operation accordingly.
+
+#### Why not using role based access control model
```

Comment:
I don't think 'roles' are a fundamental problem, I think the problem is calculating what permissions a user has, with the current system there are no APIs do to this making the only avenue to attempt to perform an action and check for 403 (bad user experience).

Might want to leave the door open in the design to allow permissions to be source from other logical groups such as roles that might be useful in the future for larger numbers of workspace users.

IMO I think this problem is mitigated if you could see a projection of an individual's users permissions, e.g. potential way to see permissions where you can get a full list, and that list has references to the source of the permission:
```
GET /user/{USER_ID}/_permissions
{
  "user": "{USER_ID}",
  "permissions": [
    {
      "name": "WORKSPACE.VIEW",
      "scope": "*",
      "from": {
        "source": "role",
        "roleName": "AllUsers",
        "ref": "/roles/AllUsers"
      }
    },
    {
      "name": "WORKSPACE.EDIT",
      "scope": "index",
      "from": {
        "source": "user",
        "ref": "/user/{USER_ID}"
      }
    }
  ]
}
```